src/Clay/Grid.hs

```diff
@@ -1,58 +1,271 @@
+gridAutoRows = key "grid-auto-rows"
+
+-- | Property defines the line names and track sizing functions of the grid columns.
+gridAutoColumns :: GridAutoTrackList a -> Css
+gridAutoColumns = key "grid-auto-columns"
+
+newtype GridAutoTrackList a = GridAutoTrackList Value
+  deriving (Val, Auto, MinContent, MaxContent, Inherit, Initial, Unset)
+
+instance IsList (GridAutoTrackList a) where
+  type Item (GridAutoTrackList a) = Size a
+  toList = error ""
+  fromList = GridAutoTrackList . noCommas
+
+-------------------------------------------------------------------------------
+gridAutoFlow :: GridAutoFlow -> Css
+gridAutoFlow = key "grid-auto-flow"
+
+newtype GridAutoFlow = GridAutoFlow Value
+  deriving (Val, Row, Column, Inherit, Initial, Unset)
+
+dense :: GridAutoFlow
+dense = GridAutoFlow "dense"
+
+rowDense :: GridAutoFlow
+rowDense = GridAutoFlow "row dense"
+
+columnDense :: GridAutoFlow
+columnDense = GridAutoFlow "column dense"
+-------------------------------------------------------------------------------
+-- | Property defines the element location inside grid template
+gridArea :: GridArea -> Css
+gridArea = key "grid-area"
+
+blankGridArea :: GridArea
+blankGridArea = GridArea "."
+
+newtype GridArea = GridArea Text
+  deriving (IsString, Val)
+
+-------------------------------------------------------------------------------
+
+gridRowStart :: GridLocation -> Css
+gridRowStart = key "grid-row-start"
+
+gridRowEnd :: GridLocation -> Css
+gridRowEnd = key "grid-row-end"
+
+gridColumnStart :: GridLocation -> Css
+gridColumnStart = key "grid-column-start"
+
+gridColumnEnd :: GridLocation -> Css
+gridColumnEnd = key "grid-column-end"
+
+gridLocation :: IsSpan -> These Integer GridArea -> GridLocation
+gridLocation isSpan = GridLocation_Data . GridLocationData isSpan
+
+data IsSpan = Span | NoSpan
+  deriving (Show, Eq)
+
+data GridLocation
+  = GridLocation_Keyword Value
+  | GridLocation_Data GridLocationData
+
+instance Val GridLocation where
+  value (GridLocation_Keyword v) = v
+  value (GridLocation_Data d) = value d
+
+instance Auto    GridLocation where auto    = GridLocation_Keyword auto
+instance Inherit GridLocation where inherit = GridLocation_Keyword inherit
+instance Initial GridLocation where initial = GridLocation_Keyword initial
+instance Unset   GridLocation where unset   = GridLocation_Keyword unset
+
+-- See under syntax:
+-- https://developer.mozilla.org/en-US/docs/Web/CSS/grid-column-start
+-- either grid index and/or named grid area is required, but span is optional
+data GridLocationData = GridLocationData IsSpan (These Integer GridArea)
+
+instance Val GridLocationData where
+  value (GridLocationData isSpan indexAndOrGridArea) =
+    if isSpan == Span
+    then value (span :: Value, indexAndOrGridArea)
+    else value indexAndOrGridArea
+
+pattern GridIndex :: Integer -> GridLocation
+pattern GridIndex n = GridLocation_Data (GridLocationData NoSpan (This n))
+
+instance Num GridLocation where
```

Comment:
There should not be a `Num` instance if most of the methods are `error`. Rather give the two methods you're interested in separate functions.

Reply:
In this particular case, it seems that a `Num` instance is confusing because many values for `GridLocation` aren't numbers. Right?